src/SessionFactory.php

```diff
@@ -0,0 +1,29 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-session for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive-session/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Expressive\Session;
+
+use Psr\Container\ContainerInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Diactoros\ServerRequestFactory;
+
+class SessionFactory
+{
+    public function __invoke(ContainerInterface $container) : SessionInterface
+    {
+        if ($container->has(SessionInterface::class)) {
+            return $container->get(SessionInterface::class);
+        }
+
+        $request = $container->has(ServerRequestInterface::class) ? $container->get(ServerRequestInterface::class) : ServerRequestFactory::fromGlobals();
```

Comment:
This is where things are going seriously wrong: a `ServerRequestInterface` is NOT a service.

This will break for long running processes.

Also the session is NOT a service - it is bound to the current request (and specifically some data in the request, usually an identifier).

Also see https://igor.io/2013/03/31/stateless-services.html

PIease don't attempt to build this stuff in this direction.

Reply:
Sure, you can remove the `Request` from constructor. Then hack on to move this to middleware and use a method `setRequest` on `LazySession` as https://github.com/zendframework/zend-expressive-session/pull/12/files#diff-b7d2bc2f8fe95a544a16ecd1a3db0c9bL32 . Does that make it stateless ?

Reply:
> The container calls setRequest on services whenever the request scope changes. This is even worse than scopes.

@Ocramius  from the article you have pointed it looks there is no solutions for the problem. Even though it recommends some of the ways, it doesn't fit into my current needs or in expressive.

I don't know if there is a way to solve this issue though.

Reply:
Not in this way - you'd simply move the session to a parameter. If you want to look at examples of how this is done cleanly, check https://github.com/psr7-sessions/storageless for a prototype/idea of the architecturally correct approach.

This is really just functional concepts applied to sessions.

Reply:
@Ocramius can you directly point me to the example you are mentioning ?  If it is via https://github.com/psr7-sessions/storageless/blob/e9d5b23503e68b10d19ac637731fef86ebb9f4ca/examples/index.php#L55 then I can't do that for the current time. I may need to change at many places.

Reply:
> I may need to change at many places.

Precisely